wait_for: change file check to handle directories and sockets

Ferroin opened the pull request:
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
wait_for

##### ANSIBLE VERSION
Based on: 8d5f4dda06d876a380f25afc2958f951544ff4ca
Originally discovered with:
```
ansible 2.3.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
open(path) throws an error when called on a directory or UNIX socket,
and therefore a check to ensure that the path is absent will always
succeed when there is a directory or file located there.

This updates the check to use os.access(path, os.F_OK) instead, which
instead just checks that the path exists instead of trying to open it as
a file, and therefore properly handles directories and sockets.

This causes a slight semantic change in how permissions are handled.
The existing code will fail to work correctly if the user running the
module on the managed host has no read access to the path specified.
The new code will work correctly in that situation.
Both versions fail if the user can't traverse the parent directory.

Fixes: #20870 

I've also added a check to the try block to catch OSError.  I've seen this call fail with an OSError on rare occasion in the face of odd extended permissions (usually MAC configuration) in cases where it should technically return False.  In such cases, the file is functionally inaccessible to the user making the call, so it's essentially not there, but it can't be created by them either.  This particular issue also technically affected the original code too, and should probably be documented somewhere.  I can make a pull request to add this to the documentation if desired.

gregswift commented:
nice and simple fix. 👍 

If you want to update the doc in the module with that information that would be good. that shoud fit inside the scope of this PR.

Ferroin commented:
OK, I'll look at getting that done before the weekend.  Would you prefer two separate commits or would you like me to just squash that into the existing commit?

Ferroin committed:
wait_for: change file check to handle directories and sockets

open(path) throws an error when called on a directory or UNIX socket,
and therefore a check to ensure that the path is absent will always
succeed when there is a directory or file located there.

This updates the check to use os.access(path, os.F_OK) instead, which
instead just checks that the path exists instead of trying to open it as
a file, and therefore properly handles directories and sockets.

This causes a slight semantic change in how permissions are handled.
The existing code will fail to work correctly if the user running the
module on the managed host has no read access to the path specified.
The new code will work correctly in that situation.
Both versions fail if the user can't traverse the parent directory.

I've also added a check to the try block to catch OSError. I've seen
this call fail with an OSError on rare occasion in the face of odd
extended permissions (usually MAC configuration) in cases where it
should technically return False. In such cases, the file is functionally
inaccessible to the user making the call, so it's essentially not there,
but it can't be created by them either. I've documented this, as well as
the fact that the bug this change fixes exists, and a rather nasty
inconsistency involving symbloic handling that I stumbled across while
testing this change.

Fixes: #20870

Ferroin commented:
I just updated after rebasing and adding the documentation.  I've got it listing version 2.4 as the first that fixes this because I assume that will be the next release, if I'm wrong about it feel free to change it.

I also added a note about this module following symbolic links when checking paths.  This caught me by surprise when I was testing (most of the other modules seem to not follow symbolic links), so I figured it should probably be documented too.

gregswift commented:
shipit

gregswift commented:
shipit

Ferroin commented:
Bumping this since it hasn't gone anywhere in a week.

gregswift commented:
bot_status

gregswift commented:
@Ferroin needs more shipits because its supported by core.

```core: Members of the Ansible organization typically do all the maintainence on this module, so only they can approve changes. Expect reviews to take longer than most other modules because of the volume the core team has on a daily basis.```

Ferroin commented:
Ah, I had forgotten that this was part of the core modules.

Ferroin commented:
Bumping this again in the hope that I won't have to continue running a locally patched version for the foreseeable future...

gregswift commented:
@bcoca (who else should i prod?) can we check this out? i already shipit'd it but with the merge of the git repos and this being a core module i'm not special enough anymore :D

gundalow commented:
Rebase needed

Ferroin commented:
And of course I forgot to fix the PEP8 errors when rebasing...

gregswift commented on lib/ansible/modules/utilities/logic/wait_for.py:
```diff
@@ -179,6 +185,7 @@
 import socket
 import sys
 import time
+import os
```
You'll want to rebase again i think some changes where just made around this per #25549

Ferroin replied:
Thanks for the heads up about that.

Ferroin commented:
ready_for_review

gregswift commented:
ship_it

AnderEnder commented:
@Ferroin, there is a merge conflict

Ferroin commented:
Just rebased and fixed the conflict.

Ferroin commented:
Poking this yet again, no merge conflicts and all checks are passing...

Also, it looks like the bot may have gotten confused, the needs_revision tag added on 2017-01-09 never got removed despite the comment that prompted it being addressed.

gregswift commented:
@Ferroin so wednesdays are meeting day for ansible community. maybe if you can make the 2pn (CDT) working group meeting you could get some more traction? #ansible-meeting on freenode.

lae commented:
I just happened to hit the bug this fixes. Any chance of getting this merged soon?

hvhaugwitz commented:
I'm also affected by this issue. Please consider to merge this pull request.

Ferroin commented:
@hvhaugwitz @lae  If you can confirm here that the PR fixes things for you, it might help finally get the attention of the core team members again so that it could finally get merged.

Ferroin commented:
bot_broken

The needs_revision tag from 2017-06-09 never got removed despite multiple revisions occurring since then.

lae commented:
This PR fixes the immediate issue I was facing regarding ensuring directories were absent (the builds in ansible-role-proxmox above failed because of a different underlying issue).

hvhaugwitz commented:
@Ferroin The PR fixes my issue regarding watching for the absence of files the user cannot open (due to file permissions) but see (due to directory permissions).

mattclay commented:
bot_status
